pymc3/step_methods/pgbart.py

```diff
@@ -275,6 +327,210 @@ def set_particle_to_step(self, t):
+        X[current_node.idx_data_points, selected_predictor],
+        missing_data,
+    )
+    # This can be unsuccessful when there are not available splitting rules
+    if available_splitting_rules.size == 0:
+        return False, None
+
+    index_selected_splitting_rule = discrete_uniform_sampler(len(available_splitting_rules))
+    selected_splitting_rule = available_splitting_rules[index_selected_splitting_rule]
+    new_split_node = SplitNode(
+        index=index_leaf_node,
+        idx_split_variable=selected_predictor,
+        split_value=selected_splitting_rule,
+    )
+
+    left_node_idx_data_points, right_node_idx_data_points = get_new_idx_data_points(
+        new_split_node, current_node.idx_data_points, X
+    )
+
+    left_node_value = draw_leaf_value(
+        sum_trees_output[left_node_idx_data_points], mean, m, normal, mu_std
+    )
+    right_node_value = draw_leaf_value(
+        sum_trees_output[right_node_idx_data_points], mean, m, normal, mu_std
+    )
+
+    new_left_node = LeafNode(
+        index=current_node.get_idx_left_child(),
+        value=left_node_value,
+        idx_data_points=left_node_idx_data_points,
+    )
+    new_right_node = LeafNode(
+        index=current_node.get_idx_right_child(),
+        value=right_node_value,
+        idx_data_points=right_node_idx_data_points,
+    )
+    tree.grow_tree(index_leaf_node, new_split_node, new_left_node, new_right_node)
+
+    return True, index_selected_predictor
+
+
+def get_available_splitting_rules(X_j, missing_data):
+    if missing_data:
+        X_j = X_j[~np.isnan(X_j)]
+    values = np.unique(X_j)
+    # The last value is never available as it would leave the right subtree empty.
+    return values[:-1]
+
+
+def get_new_idx_data_points(current_split_node, idx_data_points, X):
+    idx_split_variable = current_split_node.idx_split_variable
+    split_value = current_split_node.split_value
+
+    left_idx = X[idx_data_points, idx_split_variable] <= split_value
+    left_node_idx_data_points = idx_data_points[left_idx]
+    right_node_idx_data_points = idx_data_points[~left_idx]
+
+    return left_node_idx_data_points, right_node_idx_data_points
+
+
+def draw_leaf_value(sum_trees_output_idx, mean, m, normal, mu_std):
+    """Draw Gaussian distributed leaf values"""
+    mu_mean = mean(sum_trees_output_idx) / m
+    # add cache of precomputed normal values
+    draw = normal.random() * mu_std + mu_mean
+    return draw
+
+
+def fast_mean():
+    """If available use Numba to speed up the computation of the mean."""
+    try:
+        from numba import jit
+    except ImportError:
+        return np.mean
+
+    @jit
+    def mean(a):
+        count = a.shape[0]
+        suma = 0
+        for i in range(count):
+            suma += a[i]
+        return suma / count
+
+    return mean
+
+
+def discrete_uniform_sampler(upper_value):
+    """Draw from the uniform distribution with bounds [0, upper_value)."""
+    return int(np.random.random() * upper_value)
```

Comment:
https://numpy.org/doc/stable/reference/random/generated/numpy.random.randint.html

Seems to do this

Reply:
yes, but faster. I can add a comment to the docstring.